### PR TITLE
Add SortingNetwork implementation and test

### PR DESCRIPTION
**Pull request by makaimann**
This PR adds a symbolic boolean sorting network implementation to Smt-Switch

### CODE REVIEW

**Comment by makaimann** [comment]
Yep, thanks for catching that! Fixed with cd5ddfd164e898ae9f7016a3da799e4a9dc72c5d

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:27]: But it does return a list of terms... Can you clarify the comment (in the comment)?

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:35]:
```suggestion
 *             elements of x were set to true in that model.
```

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:36]: Please consider not ignoring the edge cases :-) Maybe adding them after this explanation. Not critical though.

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:49]:
```suggestion
 *            would be [true, true, true, false]. Note that these are the model values, not the terms that will be returned from the function.
```

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:68]: Should this function be private?

Comment (yoni206) [inline include/sorting_network.h:68]: Did you consider implementing a non-recursive function?

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:77]: Can you specify what does merging mean?

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:82]: Also, make private (or protected)?

File: src/sorting_network.cpp
Comment (yoni206) [inline src/sorting_network.cpp:56]:
```suggestion
  if (len == 0 || len == 1)
```
Because I fear we might get an infinite loop if `len=0`? You check this in `sorting_network` but what if you keep dividing by 2 and then reach `0`?

File: src/sorting_network.cpp
Comment (yoni206) [inline src/sorting_network.cpp:106]: Did you consider a non-recursive version? Maybe adding an issue for this in the repo?

File: src/sorting_network.cpp
Comment (yoni206) [inline src/sorting_network.cpp:137]: Can you document what this block is doing?

File: src/sorting_network.cpp
Comment (yoni206) [inline src/sorting_network.cpp:142]: Can you assert something about `len_res_even` and `les_res_odd`? Maybe you know one if even and one is odd?

File: src/sorting_network.cpp
Comment (yoni206) [inline src/sorting_network.cpp:180]: Can you add documentation about these cases?

File: tests/test-sorting-network.cpp
Comment (yoni206) [inline tests/test-sorting-network.cpp:101]: why can't we test this with the generic solver?

File: tests/test-sorting-network.cpp
Comment (yoni206) [inline tests/test-sorting-network.cpp:51]: I'd suggest testing also an odd length , to test well the division by 2 of the pivot. Maybe even something like `3` -- because when u divide it by 2 u get `1` and then `0`, which seem like edge cases.

File: tests/test-sorting-network.cpp
Comment (yoni206) [inline tests/test-sorting-network.cpp:60]: Cool test! :)

File: tests/test-sorting-network.cpp
Comment (yoni206) [inline tests/test-sorting-network.cpp:91]: Isn't this too strong? I thought that the requirement is `>=`, not `==`.

**Review by yoni206 — CHANGES_REQUESTED** [review]

File: tests/test-sorting-network.cpp
Comment (makaimann) [reply tests/test-sorting-network.cpp:101]: I was going to post an issue about that soon. I keep getting failures from the generic solver that seem unrelated to the test. Maybe you could look into that?

File: include/sorting_network.h
Comment (makaimann) [reply include/sorting_network.h]: Yeah that was unclear, clarified in 03a6607a7a225e4b96a42cddd25fd335aed8df3d.

File: include/sorting_network.h
Comment (makaimann) [reply include/sorting_network.h]: Haha, added with 6fee1c02d9074ff5accb4d3094d3f37eb1cba863

File: include/sorting_network.h
Comment (makaimann) [reply include/sorting_network.h]: Good point, done in 6fee1c02d9074ff5accb4d3094d3f37eb1cba863.

File: src/sorting_network.cpp
Comment (makaimann) [reply src/sorting_network.cpp]: Not the best documentation, but a little better done with d6bc10dfc4f2f1fba66463bb0187347524396e1f.

File: tests/test-sorting-network.cpp
Comment (makaimann) [reply tests/test-sorting-network.cpp]: Good point, added two others to cover all cases: d6bc10dfc4f2f1fba66463bb0187347524396e1f

File: tests/test-sorting-network.cpp
Comment (makaimann) [reply tests/test-sorting-network.cpp:93]: Right, except that I'm also asserting it's not greater than. So if `y` is the sorted network, then I'm asserting `y_i` and `not y_{i+1}` which means `>=i` but not `>i`, i.e. exactly `i`.

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:27]:
```suggestion
 *          many terms are assigned to true symbolically.
```

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:23]:
```suggestion
 *         Implements a symbolic sorting network for boolean-sorted terms.
```

File: include/sorting_network.h
Comment (yoni206) [inline include/sorting_network.h:12]:
```suggestion
** \brief Implements a symbolic sorting network for boolean-sorted terms.
```

File: include/sorting_network.h
Comment (yoni206) [reply include/sorting_network.h:76]: or maybe at least mention it is a helper for sorting_network_rec?

File: include/sorting_network.h
Comment (yoni206) [reply include/sorting_network.h:81]: ping

**Review by yoni206 — APPROVED** [review]
Thanks for all the changes! For your consideration before merging, I've made some more minor comments inline.

**Comment by makaimann** [comment]
Great, thanks for the detailed review and lots of good catches! I committed your suggestions. I'll wait for CI before merging.

**Pull request merged**
